scantool/diag_l0_elm.c

```diff
@@ -1090,10 +1104,36 @@ elm_recv(struct diag_l0_device *dl0d,
 }
 
 
-/* TODO : set new wakeup message, if possible */
+/* set new wakeup message, if possible */
 static int elm_setwm(struct diag_l0_device *dl0d, struct diag_msg *pmsg) {
-	(void) dl0d;
-	(void) pmsg;
+	struct elm_device *dev;
+
+	dev = (struct elm_device *)dl0d->l0_int;
+
+	if (dev->elmflags & ELM_INITDONE) {
+		/*
+		 * It's possible to change the wakeup message after init,
+		 * but we don't currently implement this.
+		 */
+		fprintf(stderr, FLFMT "elm_setwm: tried to set wakeup message after init\n", FL);
+		return diag_iseterr(DIAG_ERR_GENERAL);
+	}
+
+	if (dev->elmflags & ELM_323_BASIC) {
+		fprintf(stderr, FLFMT "elm_setwm: ELM323 doesn't support setting wakeup message\n", FL);
+		return diag_iseterr(DIAG_ERR_GENERAL);
+	}
+
+	if (pmsg->len != 4) {
+		fprintf(stderr, FLFMT "elm_setwm: invalid message length %d\n", FL, pmsg->len);
+		return diag_iseterr(DIAG_ERR_GENERAL);
+	}
+
+	if (dev->wm != NULL)
+		diag_freemsg(dev->wm);
+	dev->wm = diag_dupsinglemsg(pmsg);
```

Comment:
Heh, I thought you would use a uint8_t wm[4] buffer in the device struct instead of doing a dup + free. Of course this way is more generic.